phac-nml/irida-next
# Project history view

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in a `History` page for projects which displays the record level changes made to the project (name, description, path, parent). The versions are displayed in a timeline view and clicking on the version number displays a modal with which attributes were changed and by whom.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

![image](https://github.com/phac-nml/irida-next/assets/25466211/a14108ca-3122-4694-b082-c6118df8a99d)

![image](https://github.com/phac-nml/irida-next/assets/25466211/6f894ed6-2aa7-487f-b024-2b446598ca0d)

![image](https://github.com/phac-nml/irida-next/assets/25466211/b45fedd8-59c6-4a29-9750-14dc071511ff)


## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Login as a user
2. Create a project
3. Go to the `History` page for the project 
4. You should see 1 version, and clicking the `Version 1` should launch a modal that displays the attributes which the project was created with
5. Go to the project settings page. Change the name
6. On the `History` page for the project you should now see 2 versions. Clicking the `Version 2` should launch a modal which displays the change to the name attribute. On the left hand side is the previous version of that attribute and on the right is the current version
7. Try the same for description, path, transferring the project to a different namespace
8. Also try to view history for other projects that were seeded that the user has direct membership or inherited membership on (shared group members should not be able to see the history for a project)

**Note:** In `rails console` you can get the log_data for a project with `.reload_log_data` on the project namespace.

Log data is in the format

```#<Logidze::History:0x00007f19a961ad18
 @data=
  {"h"=>
    [{"c"=>{"id"=>13, "name"=>"Outbreak 2022", "path"=>"outbreak-2022", "type"=>"Project", "owner_id"=>1, "parent_id"=>12, "deleted_at"=>nil, "description"=>"This is a description for project Outbreak 2022"}, "v"=>1, "ts"=>1707331399281},
     {"c"=>{"description"=>"This is a description for project Outbreak 2022a"}, "m"=>{"_r"=>1}, "v"=>2, "ts"=>1707333461429}],
   "v"=>2}>
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
